### PR TITLE
Bug 2014420: dont crash topology page

### PR DESCRIPTION
**Pull request by yaacov**
The `referenceForModel` method crash if called with `model` undefined, this may happen in the topology page.
This PR let it fail gracefully.

Screenshots:
Before:
screenshot-localhost_9000-2021 10 19-21_06_18

After:
screenshot-localhost_9000-2021 10 19-21_07_41

### CODE REVIEW

**Comment by yaacov** [comment]
@spadgett ping

**Comment by yaacov** [comment]
@spadgett ping

**Comment by yaacov** [comment]
@spadgett ping

**Comment by invincibleJai** [comment]
@yaacov I feel even this issue is connected https://bugzilla.redhat.com/show_bug.cgi?id=2020904 the root cause is the model not being discovered in the case of VMs with ApiDiscovery ... we need to look at why is this not getting discovered I think 

@rohitkrai03 @glekner any inputs?

these are contributed vis dynamic extensions here 

https://github.com/openshift/console/blob/release-4.9/frontend/packages/kubevirt-plugin/console-extensions.json#L512-L526

cc @spadgett @christianvogt

**Comment by yaacov** [comment]
@invincibleJai hi, dynamic plugins are loaded at runtime and may be broken, as a dev-console developer  I don't want to be warry about 3rd party code braking the topology view. IMHO the dev-console should fail gracefully in this cases, currently we let dynamic plugins to break the console for all users (even those who don't use this plugin)

**Comment by invincibleJai** [comment]
> @invincibleJai hi, dynamic plugins are loaded at runtime and may be broken, as a dev-console developer I don't want to be warry about 3rd party code braking the topology view. IMHO the dev-console should fail gracefully in this cases, currently we let dynamic plugins to break the console for all users (even those who don't use this plugin)

yeah we should have checks and user should not see failures in topology so your changes in this PR is needed but I was trying to highlight the other issue logged here https://bugzilla.redhat.com/show_bug.cgi?id=2020904 we need dynamic discovery for the model here which is not happening looks like.

**Comment by yaacov** [comment]
@invincibleJai yes I agree, btw/ I re-installed my cluster and the bug disappeared ... I currently don't have a cluster to reproduce  and  test the fix.

**Comment by yaacov** [comment]
@spadgett please re-review

**Comment by yaacov** [comment]
@spadgett please re review

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx
Comment (spadgett) [inline frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:33]:
```suggestion
  if (model.version) {
```

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx
Comment (spadgett) [inline frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: Why aren't we using the API version?

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx
Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx]: Thanks

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx
Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: I don't know how to get the API version at this point, `modelForGroupKind` is using the `store` to get it.
@spadgett what am I missing ?

Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: @spadgett help, what am I missing ?

Comment (invincibleJai) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: @spadgett these models are fetched form extension contributed via diff packages for topology

https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts#L36-L54

where resources have the model with type `ExtensionK8sGroupKindModel`, where version is optional but kind group is required.

https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/api/common-types.ts#L21-L25

that is why `modelForGroupKind` is used here to get models with group and kind to get models and existing logic as well does that

https://github.com/openshift/console/blob/master/frontend/packages/topology/src/data-transforms/DataModelProvider.tsx#L37

Comment (spadgett) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: Hi @invincibleJai. Thanks. I'm not familiar with the topology extension points, but isn't the API version important if the extension needs to work with the data? Why was it left out?

We can make this a follow up discussion to unblock the bug fix.

Comment (invincibleJai) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: I am not very sure about the thought process while designing these extension points for topology but what I could think of is the reason to have this flexibility would be in topology we visualize resources as diff shapes uniquely by group and kind. Packages will provide a list of resources with the group and kind we topology is interested in, as in the case of operators at times they support multiple versions for the same resource. So based on group and kind can fetch the model for the preferred version if CRD has one or the latest one with `modelForGroupKind`. 

cc @christianvogt  can add more to it.

Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: > We can make this a follow up discussion to unblock the bug fix.

This bug affect users, unblocking the bug and creating a follow up discussion makes a lot of sense to me too.

Comment (invincibleJai) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: Thanks Sam, have logged one here https://bugzilla.redhat.com/show_bug.cgi?id=2039381

@yaacov based on the original ticket, have created a new one but to reproduce team would need help to get `kubevirt CI cluster` and even the issue was intermittent https://github.com/openshift/console/pull/10272#issuecomment-962907048 

so if there are steps to reproduce consistently or any way to reproduce it on another cluster, pls add a comment on the ticket.

Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: thanks, added @gouyang to the new bugzilla.

Comment (yaacov) [reply frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:39]: > OK I don't want to block the bug any longer. Let's go ahead with the fix. Can we open a bug to track the follow on issue?

@spadgett hi, do I need to change something before you can lgtm it ?

**Review by spadgett — APPROVED** [review]
/lgtm

**Pull request merged**
